### PR TITLE
[DPP-1144] Retrieving a Daml packageId from Ledger API server for Benchtool

### PR DESCRIPTION
**Pull request by pbatko-da**
changelog_begin
changelog_end

### CODE REVIEW

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      } yield {
+        val packageId = getPackageResponse.hash
+        val packageName = decodePackageName(
+          archivePayloadBytes = getPackageResponse.archivePayload.toByteArray,
+          pkgId = Ref.PackageId.assertFromString(packageId),
+        )
+        packageName -> packageId
+      }
+      val candidatesPackageIds =
+        packageNamesToPackageIds.filter(_._1 == BenchtoolTestsPackageName).map(_._2)
```
Comment (skisel-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala:78]: collect?

**Review by skisel-da — APPROVED** [review]

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala (hunk 1 above)
Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala]: Fixed

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        sys.error(s"Could not find a Daml package with name '$BenchtoolTestsPackageName'")
+      )
+      BenchtoolTestsPackageInfo(packageId = packageId)
+    }
+  }
+
+  private def decodePackageName(archivePayloadBytes: Array[Byte], pkgId: Ref.PackageId): String = {
+    val pkg: Ast.Package = ArchivePayloadParser
+      .andThen(Reader.readArchivePayload(pkgId, _))
+      .andThen(Decode.decodeArchivePayload(_))
```
Comment (remyhaemmerle-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala:92]:
```suggestion
    val pkg: Ast.Package = com.daml.lf.archive.archivePayloadDecoder(pkgId, onlySerializableDataDefs = false).
```

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala [hunk 3]
```diff
@@ -38,7 +38,7 @@ final class GenReader[X] private[archive] (
   def fromFile(file: java.io.File): Either[Error, X] =
     fromFile(file.toPath)
 
-  private[archive] def andThen[Y](f: X => Either[Error, Y]): GenReader[Y] =
+  def andThen[Y](f: X => Either[Error, Y]): GenReader[Y] =
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala:41]: Please revert this change.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/** Contains utilities for retrieving useful facts
+  * from data already submitted to a Ledger API server.
+  * (The motivating use case are the benchmarks that do not perform a submission step on their own
+  * and for that reason cannot statically determine these facts.)
+  */
+object SubmittedDataAnalyzing {
```
Comment (simonmaxen-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala:27]: How about just `SubmittedData` ?

Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala:27]: But this class is no the submitted data; so maybe `SubmittedDataUtils`?

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala (hunk 2 above)
Comment (pbatko-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/SubmittedDataAnalyzing.scala]: Ok

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala (hunk 3 above)
Comment (pbatko-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/GenReader.scala]: Ok
